src/workerd/api/trace.h

```diff
@@ -104,181 +118,195 @@ class TraceItem::FetchEventInfo final: public jsg::Object {
   class Request;
   class Response;
 
-  explicit FetchEventInfo(kj::Own<Trace> trace, const Trace::FetchEventInfo& eventInfo,
+  explicit FetchEventInfo(jsg::Lock& js,
+                          const Trace& trace,
+                          const Trace::FetchEventInfo& eventInfo,
                           kj::Maybe<const Trace::FetchResponseInfo&> responseInfo);
 
   jsg::Ref<Request> getRequest();
   jsg::Optional<jsg::Ref<Response>> getResponse();
 
   // TODO(cleanup) Use struct types more?
   JSG_RESOURCE_TYPE(FetchEventInfo) {
-    JSG_READONLY_INSTANCE_PROPERTY(response, getResponse);
-    JSG_READONLY_INSTANCE_PROPERTY(request, getRequest);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(response, getResponse);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(request, getRequest);
   }
 
 private:
-  kj::Own<Trace> trace;
-  const Trace::FetchEventInfo& eventInfo;
-  kj::Maybe<const Trace::FetchResponseInfo&> responseInfo;
+  jsg::Ref<Request> request;
+  jsg::Optional<jsg::Ref<Response>> response;
 };
 
 class TraceItem::FetchEventInfo::Request final: public jsg::Object {
 public:
-  explicit Request(kj::Own<Trace> trace, const Trace::FetchEventInfo& eventInfo);
-
-  jsg::Optional<v8::Local<v8::Object>> getCf(v8::Isolate* isolate);
+  struct Detail : public kj::Refcounted {
+    jsg::Optional<jsg::V8Ref<v8::Object>> cf;
+    kj::Array<Trace::FetchEventInfo::Header> headers;
+    kj::String method;
+    kj::String url;
+
+    Detail(jsg::Optional<jsg::V8Ref<v8::Object>> cf,
+           kj::Array<Trace::FetchEventInfo::Header> headers,
+           kj::String method,
+           kj::String url);
+  };
+
+  explicit Request(jsg::Lock& js, const Trace& trace, const Trace::FetchEventInfo& eventInfo);
+  explicit Request(Detail& detail);
+  // Creates an unredacted instance that shared a ref of the Detail
+
+  jsg::Optional<jsg::V8Ref<v8::Object>> getCf(jsg::Lock& js);
   jsg::Dict<jsg::ByteString, jsg::ByteString> getHeaders();
   kj::StringPtr getMethod();
   kj::String getUrl();
 
   jsg::Ref<Request> getUnredacted();
 
   JSG_RESOURCE_TYPE(Request) {
-    JSG_READONLY_INSTANCE_PROPERTY(cf, getCf);
-    JSG_READONLY_INSTANCE_PROPERTY(headers, getHeaders);
-    JSG_READONLY_INSTANCE_PROPERTY(method, getMethod);
-    JSG_READONLY_INSTANCE_PROPERTY(url, getUrl);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(cf, getCf);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(headers, getHeaders);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(method, getMethod);
+    JSG_LAZY_READONLY_INSTANCE_PROPERTY(url, getUrl);
 
     JSG_METHOD(getUnredacted);
   }
 
 private:
-  kj::Own<Trace> trace;
-  const Trace::FetchEventInfo& eventInfo;
   bool redacted = true;
+  kj::Own<Detail> detail;
```

Comment:
It looks like here we are replacing a refcounted `kj::Own<Trace>` with a refcounted `kj::Own<Detail>`.  Am I correct in assuming that the difference is that we are ensuring that all the refcounted consumers of Detail are in a single isolate, such that this code does not have the same problem with multithreaded garbage collection as the code it's replacing?